src/instructlab/model/train.py

```diff
@@ -941,54 +988,158 @@ def _run_phased_training(
     phase2_eval_cache: pathlib.Path,
     mtbench_judge: pathlib.Path,
     enable_serving_output: bool,
+    journal: TrainingJournal,
 ) -> None:
-    click.secho("Training Phase 1/2...", fg="cyan")
-
-    _training_phase(
-        train_args=train_args.model_copy(deep=True),
-        torch_args=torch_args,
-        data_path=phase1_data,
-        checkpoint_dir=phase1_checkpoints_dir,
-        num_epochs=phase1_num_epochs,
-        samples_per_save=phase1_samples_per_save,
-        effective_batch_size=phased_phase1_effective_batch_size,
-        # model override not necessary because we expect model to come from ctx.params.model_path.
-    )
+    logger.debug("Starting multi-phase training.\n%s", journal.print_model_rich())
+    if journal.current_phase == TrainingPhases.DONE:
+        click.secho(
+            "The selected Training Journal suggests that training has finished, with 'current_phase=done' in the journalfile.",
+            fg="cyan",
+        )
+        return
+
+    # make mypy happy
+    phase_model: TrainPhaseModel | EvalPhaseModel | None = None
+
+    if journal.current_phase == TrainingPhases.TRAIN1:
+        click.secho("Training Phase 1/2...", fg="cyan")
+
+        phase_model = journal.journal.train_1
+        if phase_model is None:
+            phase_model = TrainPhaseModel(checkpoints=phase1_checkpoints_dir)
+            journal.journal.train_1 = phase_model
+        journal.commit()
+
+        _training_phase(
+            train_args=train_args.model_copy(deep=True),
+            torch_args=torch_args,
+            data_path=phase1_data,
+            checkpoint_dir=phase1_checkpoints_dir,
+            num_epochs=phase1_num_epochs,
+            samples_per_save=phase1_samples_per_save,
+            effective_batch_size=phased_phase1_effective_batch_size,
+            # model override not necessary because we expect model to come from ctx.params.model_path.
+        )
 
-    click.secho("MMLU evaluation for Phase 1...", fg="cyan")
-    # NOTE: requires hf_format sub-directory. Training sets this up.
-    phase1_checkpoints_dir = phase1_checkpoints_dir / "hf_format"
-    _, p1_best_ckpt = _evaluate_dir_of_checkpoints(
-        checkpoints_dir=phase1_checkpoints_dir, eval_func=_mmlu
-    )
+        phase_model.ended_at_utc = TrainingJournal.now_utc()
 
-    click.secho("Training Phase 2/2...", fg="cyan")
-    _training_phase(
-        train_args=train_args.model_copy(deep=True),
-        torch_args=torch_args,
-        data_path=phase2_data,
-        checkpoint_dir=phase2_checkpoints_dir,
-        model_override=p1_best_ckpt,
-        num_epochs=phase2_num_epochs,
-        samples_per_save=phase2_samples_per_save,
-        effective_batch_size=phased_phase2_effective_batch_size,
-    )
+        journal.current_phase = TrainingPhases.EVAL1
+        journal.commit()
```

Comment:
We should definitely have a context manager for this `journal.commit` so that we can fail graciously in case of error and commit otherwise.

Reply:
What do you mean? If there's an error, _training_phase will raise an exception and exit, not making it to the commit statement.